### PR TITLE
Mainly changes to kotlin.stg to fix various issues.  Added Unit test …

### PR DESCRIPTION
**Pull request by OldWartHog**
…but can't get gradle to actually execute it.  Changes to kotlin.stg as follows:

  * BaseListener class needs to be 'open'.
  * Changed Rule class definitions to be more Kotlin like.  (Really to stop IDE complaining so much about it)
  * In AltBlock, (line 520) added '{}' around body of when clause, doesn't compile without this.
  * In OptionalBlock (line 529), removed incorrect ":"
  * TokenPropertyRef's ad RulePropertyRef's where still outputting Java syntax.
  * List<TerminalNode> incorrectly cast as TerminalNode (line 710)
  * 'i' sometimes declared as (int i) rather than (I: Int) (line 712)
  * Changed ParserRuleContext class definition to be more Kotlin-like (lines 741 & 747).  Not strictly necessary but improves output code.
  * Attribute declarations didn't work, made output more kotlin, but still not correct.
  * Sometimes when clause used syntax: ': ->' rather than '->'

### CODE REVIEW

**Comment by ftomassetti** [comment]
Hi @OldWartHog , thank you so much for your contribution!
It seems there is an issue with failing test: some generated code is not compiling. You should be able to read here about the error:

https://travis-ci.org/ftomassetti/antlr-kotlin/builds/392625221?utm_source=github_status&utm_medium=notification

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg [hunk 1]
```diff
@@ -120,7 +120,7 @@ import org.antlr.v4.kotlinruntime.tree.TerminalNode
  * which can be extended to create a listener which only needs to handle a subset
  * of the available methods.
  */
-class <file.grammarName>BaseListener : <file.grammarName>Listener {
+open class <file.grammarName>BaseListener : <file.grammarName>Listener {
```
Comment (ftomassetti) [inline antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg:123]: Right!

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg [hunk 2]
```diff
@@ -371,7 +371,7 @@ RuleFunction(currentRule,args,code,locals,ruleCtx,altLabelCtxs,namedActions,fina
 <altLabelCtxs:{l | <altLabelCtxs.(l)>}; separator="\n">
 
 fun <if(currentRule.modifiers)><currentRule.modifiers:{f | <f> }><else><endif> <currentRule.name>(<args; separator=",">) : <currentRule.ctxType> {
-	var _localctx : <currentRule.ctxType> = <currentRule.ctxType>(context, state<currentRule.args:{a | , <a.name>}>)
+	val _localctx : <currentRule.ctxType> = <currentRule.ctxType>(context, state<currentRule.args:{a | , <a.name>}>)
```
Comment (ftomassetti) [inline antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg]: Goog idea

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg [hunk 3]
```diff
@@ -707,9 +707,9 @@ RuleContextListDecl(rdecl) ::= "var <rdecl.name> : MutableList\<<rdecl.ctxName>>
 ContextTokenGetterDecl(t)      ::=
     "fun <t.name>() : TerminalNode = getToken(<parser.name>.Tokens.<t.name>.id, 0) as TerminalNode"
 ContextTokenListGetterDecl(t)  ::=
-    "fun <t.name>() : List\<TerminalNode> = getTokens(<parser.name>.Tokens.<t.name>.id) as TerminalNode"
+    "fun <t.name>() : List\<TerminalNode> = getTokens(<parser.name>.Tokens.<t.name>.id)"
 ContextTokenListIndexedGetterDecl(t)  ::= <<
-fun <t.name>(int i) : TerminalNode = getToken(<parser.name>.Tokens.<t.name>.id, i) as TerminalNode
+fun <t.name>(i: Int) : TerminalNode = getToken(<parser.name>.Tokens.<t.name>.id, i) as TerminalNode
```
Comment (ftomassetti) [inline antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg:712]: Good catch

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg [hunk 4]
```diff
@@ -738,16 +738,15 @@ CaptureNextTokenType(d) ::= "<d.varName> = _input!!.LA(1)"
 
 StructDecl(struct,ctorAttrs,attrs,getters,dispatchMethods,interfaces,extensionMembers)
 	::= <<
-open class <struct.name> : <if(contextSuperClass)><contextSuperClass><else>ParserRuleContext<endif><if(interfaces)> : <interfaces; separator=", "><endif> {
+open class <struct.name>(parent: ParserRuleContext?, invokingState: Int) : <if(contextSuperClass)><contextSuperClass><else>ParserRuleContext<endif>(parent, invokingState)<if(interfaces)> : <interfaces; separator=", "><endif> {
```
Comment (ftomassetti) [inline antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg]: Maybe it is here the problem that is causing the compilation issue. It involves some constructor not being called

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg [hunk 5]
```diff
@@ -738,16 +738,15 @@ CaptureNextTokenType(d) ::= "<d.varName> = _input!!.LA(1)"
...
     override var ruleIndex: Int
         get() = Rules.RULE_<struct.derivedFromName>.id
         set(value) { throw RuntimeException() }
 	<attrs:{a | <a>}; separator="\n">
 	<getters:{g | <g>}; separator="\n">
-	<if(ctorAttrs)>constructor(parent: ParserRuleContext?, invokingState: Int) : super(parent, invokingState){ ; }<endif>
-	constructor(parent: ParserRuleContext?, invokingState: Int<ctorAttrs:{a | , <a>}>) : super(parent, invokingState){
-		<struct.ctorAttrs:{a | this.<a.name> = <a.name>;}; separator="\n">
-	}
+	<if(ctorAttrs)>constructor(parent: ParserRuleContext?, invokingState: Int<ctorAttrs:{a | , <a>}>) : this(parent, invokingState){
```
Comment (ftomassetti) [inline antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg]: Don't we always need to call the constructor with the extra params now?

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg (hunk 4 above)
Comment (ftomassetti) [reply antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg]: Why do we need to always pass these two params to the constructor?

**Review by ftomassetti — COMMENTED** [review]
There is an issue with tests failing and some questions on the constructor added. The rest looks good and your work is appreciated

**Comment by ftomassetti** [comment]
The build is still failing https://travis-ci.org/ftomassetti/antlr-kotlin/builds/394034381?utm_source=github_status&utm_medium=notification

```
<=------------> 8% EXECUTING [25s]> :compileKotline/travis/build/ftomassetti/antlr-kotlin/antlr-kotlin-examples-jvm/generated-src/antlr/MiniCalcParser.kt: (594, 19): Primary constructor call expected
e: /home/travis/build/ftomassetti/antlr-kotlin/antlr-kotlin-examples-jvm/generated-src/antlr/MiniCalcParser.kt: (875, 19): Primary constructor call expected
e: /home/travis/build/ftomassetti/antlr-kotlin/antlr-kotlin-examples-jvm/generated-src/antlr/MiniCalcParser.kt: (910, 39): Val cannot be reassigned
e: /home/travis/build/ftomassetti/antlr-kotlin/antlr-kotlin-examples-jvm/generated-src/antlr/MiniCalcParser.kt: (945, 19): Primary constructor call expected
e: /home/travis/build/ftomassetti/antlr-kotlin/antlr-kotlin-examples-jvm/generated-src/antlr/MiniCalcParser.kt: (988, 28): Val cannot be reassigned
```

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg (hunk 4 above)
Comment (OldWartHog) [reply antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg]: I think I was a too quick to change this - I'm going to back out this change and the one below that goes with it.

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg (hunk 5 above)
Comment (OldWartHog) [reply antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg]: As above - I'm going to back out this change

**Comment by ftomassetti** [comment]
This stuff is tricky, it is normal it takes a few trials to get it right

**Comment by ftomassetti** [comment]
Good, the constructor issues disappeared, we have just the reassigned vals:

```
e: /home/travis/build/ftomassetti/antlr-kotlin/antlr-kotlin-examples-jvm/generated-src/antlr/MiniCalcParser.kt: (928, 39): Val cannot be reassigned
e: /home/travis/build/ftomassetti/antlr-kotlin/antlr-kotlin-examples-jvm/generated-src/antlr/MiniCalcParser.kt: (1008, 28): Val cannot be reassigned
```

**Comment by ftomassetti** [comment]
Now it seems to be compiling but one test is failing:
```
MiniCalcParserTest > callListener FAILED
    kotlin.NotImplementedError at MiniCalcParserTest.kt:116
```

**Comment by ftomassetti** [comment]
A test using the JVM target is still failing:

```
MiniCalcParserTest > callListener FAILED
    kotlin.NotImplementedError at MiniCalcParserTest.kt:116
```

**Comment by ftomassetti** [comment]
Everything passing. Great work @OldWartHog , your contribution is very appreciated!

**Pull request merged**